Changes to support hiding/showing tour description text

tjmw opened the pull request:
Splits the booking option title and description apart, and splits the description into a base and possible overflow, which the client can style/handle as they choose.

Relied on by ETMAInc/TruffleHunt-UI#223

joshuaclayton commented on src/BookingWidget/View.elm:
```diff
@@ -143,28 +146,66 @@ selectTicketCount attributes ticket =
+tourDescriptionTitle attributes translator { booking } =
+    booking.option
+        |> Maybe.andThen .description
+        |> Maybe.map (\_ -> small attributes [ text <| translator TourDescriptionHeadline ])
```
this feels weird to me; since we're discarding the `Just a` value, it seems the `Maybe`s are used here to check presence more than anything else; is that correct?

tjmw replied:
Yeah in this case that's correct (we want to render the title only if there's a description). I agree that it feels weird. I'll try out some of the suggestions made in slack and see what works well.

joshuaclayton commented on src/BookingWidget/View.elm:
```diff
@@ -143,28 +146,66 @@ selectTicketCount attributes ticket =
+renderSplitDescription : List (Attribute Msg) -> Int -> String -> Html Msg
+renderSplitDescription attributes characterLimit description =
+    let
+        ( preview, overflow ) =
```
Thoughts on pulling this out into a new type?

tjmw replied:
:+1: `StringWithOverflow` or `DescriptionWithOverFlow` perhaps?

joshuaclayton commented on src/BookingWidget/View.elm:
```diff
@@ -143,28 +146,66 @@ selectTicketCount attributes ticket =
+        ( _, overflow ) =
+            splitStringAtLimit description characterLimit
+    in
+        String.isEmpty overflow
```
With the aforementioned new type, you could then represent this more easily as well.

tjmw replied:
Right, because you could write something like `hasOverflow : StringWithOverFlow -> Bool`

joshuaclayton commented on src/BookingWidget/View.elm:
```diff
@@ -143,28 +146,66 @@ selectTicketCount attributes ticket =
+
+takeWordsUntilLimitReached : ( List String, List String ) -> Int -> ( List String, List String )
+takeWordsUntilLimitReached ( firstList, secondList ) limit =
+    case List.isEmpty secondList || String.length (String.join " " firstList) > limit of
```
Thoughts on something like this for handling splitting strings? https://ellie-app.com/3Bgs2KMwqcfa1/0

tjmw replied:
Oh yeah that looks good. This was a fun problem to figure out, and I was realising there were a bunch of possibilities to choose from.

tjmw commented:
Hi @joshuaclayton thanks again for the feedback. I introduced a `SplitString` class (like in your Ellie app example) and that cleaned things up a bunch. I also removed that questionable usage of `Maybe.map` where we were throwing away the value. I didn't change the actual string splitting logic yet (it's now in `BookingWidget.SplitString`) just so I can get moving on to other tasks. Now it's separated out that logic can be modified/refactored in isolation.

germsvel commented on src/BookingWidget/View.elm:
```diff
@@ -145,28 +149,69 @@ selectTicketCount attributes ticket =
-
-            Nothing ->
-                div [] []
+        case stringWithOverflow of
```
What do you think about moving this to the `SplitString` module? It could be something like `SplitString.isStringWithOverflow string`.

tjmw replied:
Another good question :-) I did have `hasOverflow : SplitString -> Bool` previously, but switched to the case statement because I need to do the pattern matching to pull the separate parts out anyway. I played around with `preview` and `overflow` functions which take a `SplitString` and return a `String` (or in the case of `overflow` a `Maybe String` since you don't have an overflow if there isn't one) but it was a bit messy. This `case` felt like the cleanest options I found.

germsvel replied:
Oh, I like the `case` statement. I'm not sure I communicated that well. 

I was just suggesting moving the function as is into the `SplitString` module so that the function lives there: 

```elm
isStringWithOverflow : SplitString -> Bool
isStringWithOverflow stringWithOverflow =
  case stringWithOverflow of 
    StringWithOverflow _ _ -> 
        False
    StringWithoutOverflow _ -> 
        True
```

tjmw replied:
Oh thanks for clarifying, I'd misunderstood before, my bad. Yep that sounds like a great suggestion!

tjmw replied:
Addressed in 8b309ae5d12968de753e14685315c03cc61b9b7e.

germsvel commented on src/BookingWidget/SplitString.elm:
```diff
@@ -0,0 +1,43 @@
+module BookingWidget.SplitString
```
What do you think about changing the name of this module? It wasn't clear to me when I first read this what this module was supposed to do. `SplitString` sounds like an action you're performing on a string, but it doesn't really convey that you're asking for a string and its overflow. 

Maybe something like `StringWithOverflow`? It feels like that's more what this module is all about. 

Then, your type could become something like this, 

```elm
type StringWithOverflow
  = HasOverflow String String
  | NoOverflow String
```

And then when you're using the module it would read like this, `StringWithOverflow.fromString "hello world" 1`

tjmw replied:
Good questions. I initially had `StringWithOverflow` (or something very similar) and felt a little weird about it since we don't always have an overflow, and I thought the name might be misleading. Josh used `SplitString` in the Ellie App example he linked to above, and I liked it. The good news going forward is that since this isn't leaked out into the app we're free to change here without any knock on effects.

germsvel commented on src/BookingWidget/View.elm:
```diff
@@ -145,28 +149,69 @@ selectTicketCount attributes ticket =
 -}
-tourDescription : List (Attribute Msg) -> Translator -> Model -> Html Msg
-tourDescription attributes translator model =
+tourDescription : List (Attribute Msg) -> Model -> Int -> Html Msg
```
Just from curiosity, why do we use the word `tour` here? Is a tour equivalent to a booking option?

tjmw replied:
Oh good question, I hadn't thought about it before. I didn't add the function originally so not sure where the name came from, perhaps it's how the API refers to it somewhere, or elsewhere in the business perhaps. I think I'm going to leave it as is for now, just because it has implications for the app(s) too since this function is exposed.
